Adding Expression Metrics & Metric Types to Metric Page

callum-mcdata opened the pull request:
## Description & motivation
This PR adds information to the `metrics.md` file about expression metrics. Additionally, it adds a table of supported metric types. NOTE: These metric type definitions actually live within the metrics package but I believe it is important for that information to live in the docs and not just the package README. 

This PR satisfies Issue #1638 

## Prerelease docs
If this change is related to functionality in a prerelease version of dbt (delete if not applicable):
- [X] I've added versioning components, as described in ["Versioning Docs"](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/versioningdocs.md)
- [ ] I've added a note to the prerelease version's [Migration Guide](https://github.com/dbt-labs/docs.getdbt.com/tree/current/website/docs/guides/migration/versions)

callum-mcdata committed:
changes to include expression metrics

callum-mcdata commented:
Hmm if I had to hazard a guess I would say this is probably because of some `<VersionBlock>` shenanigans. I wonder if `1.2` hadn't been defined as a version block and is failing because of it. Let me see if I can peer behind the netlify curtain

callum-mcdata committed:
Merge branch 'current'

callum-mcdata commented:
Well when a certain someone hasn't done a pull from `current` since he started, no wonder there are some issues 😅 . Just pulled in the most recent changes and I believe that should fix the issue

callum-mcdata committed:
fixes!

jasnonaz reviewed (approved):
This is fantastic work! Thanks for putting this together and powering through docusaurus issues. Non blocking question for you: Do you think we should have info about the metrics package in here / link to the metrics package. Feels like this is important information for folks who are getting into metrics.

callum-mcdata commented:
@jasnonaz that is a really good point - definitely should have some links to the metrics package, just as the metrics package should have some links to this doc so people can cross reference. Let me add an additional section linking to the package

callum-mcdata committed:
updating for link

jtcohen6 commented on website/docs/docs/building-a-dbt-project/metrics.md:
```diff
@@ -29,6 +29,8 @@ A metric is a timeseries aggregation over a <Term id="table" /> that supports ze
 
 In v1.0, dbt supports metric definitions as a new node type. Like [exposures](exposures), metrics participate in the dbt DAG and can be expressed in YAML files. By defining metrics in dbt projects, you encode crucial business logic in tested, version-controlled code. Further, you can expose these metrics definitions to downstream tooling, which drives consistency and precision in metric reporting ...
```
[Non-blocking] Thoughts on updating this language? It still refers to:
- Metrics being brand-new in v1.0 — less true now than it was last December
- Metrics being **experimental** / liable to change in new minor versions — this is still true, but we anticipate removing this "experimental" designation in a few months!

Happy for those changes to come in a subsequent PR, since it's adjacent scope

callum-mcdata replied:
I agree - opened #1690 to remember this suggestion and track it in a future PR!

jtcohen6 commented on website/docs/docs/building-a-dbt-project/metrics.md:
```diff
@@ -29,6 +29,8 @@ A metric is a timeseries aggregation over a <Term id="table" /> that supports ze
 
 In v1.0, dbt supports metric definitions as a new node type. Like [exposures](exposures), metrics participate in the dbt DAG and can be expressed in YAML files. By defining metrics in dbt projects, you encode crucial business logic in tested, version-controlled code. Further, you can expose these metrics definitions to downstream tooling, which drives consistency and precision in metric reporting ...
 
+For more information on query the metrics that are defined, please reference the [dbt_metrics package.](https://github.com/dbt-labs/dbt_metrics)
```
👍  Definitely think it's more than appropriate to prominently link to the `dbt-metrics` package from this page

[Blocking] I think I know what you're saying here, but the wording is a bit tricky. Do you mean:
- "For more information on querying metrics that are defined in your dbt project"
- "For more information on how metrics queries are defined"
?

callum-mcdata replied:
Wow, yea the language here is ... not great 😅 . I will update to:

`For more information on querying the metrics defined in your dbt project`

callum-mcdata replied:
Resolved!

jtcohen6 commented on website/docs/docs/building-a-dbt-project/metrics.md:
```diff
@@ -90,6 +92,85 @@ metrics:
+| Field       | Description                                                 | Example                         | Required? |
+|-------------|-------------------------------------------------------------|---------------------------------|-----------|
+| name        | A unique identifier for the metric                          | new_customers                   | yes       |
+| model       | The dbt model that powers this metric                       | dim_customers                   |     yes (no for `expression` metrics)    |
```
[Blocking] From a quick visual inspection, I think this is the only change in the table between v1.0/v1.1 &rarr; v1.2:
- `model` + `Required?`: `yes (no for expression metrics)`, instead of `yes` (= always) 

Is that right?

I think our `VersionBlock` logic might be fancy enough to make this work:
```suggestion
| model       | The dbt model that powers this metric                       | dim_customers                   | <VersionBlock firstVersion="1.2">yes (no for `expression` metrics)</VersionBlock><VersionBlock lastVersion="1.1">yes</VersionBlock> |
```
I tried that locally and it seemed to do the trick. Then we don't need to duplicate the full table contents for different docs versions.

jtcohen6 commented on website/docs/docs/building-a-dbt-project/metrics.md:
```diff
@@ -90,6 +92,85 @@ metrics:
+- subtractions 
+- any arbitrary calculation
+
+As long as the two+ base metrics (the metrics that comprise the `expression` metric) share the specified `time_grains` and `dimensions`, those attributes can be used in any downstream metrics macro.
```
[Non-blocking]

> As long as the two+ base metrics (the metrics that comprise the `expression` metric) share the specified `time_grains` and `dimensions`

Ah - do we have validation for this in `dbt-core`? In the `dbt-metrics` package?

callum-mcdata replied:
We have "validation" for this in the `dbt_metrics` package. The macro constructs a list of acceptable dimensions from all metrics used and then the provided dimensions are compared against that list. The macro errors out if it is an invalid dimension.

jtcohen6 commented on website/docs/docs/building-a-dbt-project/metrics.md:
```diff
@@ -90,6 +92,85 @@ metrics:
+    description: "The average revenue received per customer"
+
+    type: expression
+    sql: "{{metric('total_revenue')}} / {{metric('count_of_customers')}}"
```
[Non-blocking] Is this the syntax we'd recommend, over syntax like:
```yaml
    sql: "{{ metric('total_revenue') ~ "/" ~ metric('count_of_customers') }}"
```
(Does that alt syntax even work?)

callum-mcdata replied:
I tested the alt syntax and it didn't appear to work. Even testing a few different options around quotation placement returned unicode errors. 

To me that is fine anyway because I wouldn't want users to have to worry about concatenation within the sql field. Self contained makes most sense to me and personally feels the most ergonomic.

jtcohen6 reviewed (changes requested):
@callum-mcdata Thanks for taking this on! I left a few comments, and called out which ones are [blocking] or [non-blocking] for merging this PR.

Could you also add a note to [the "New and changed documentation" section of the v1.2 migration guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/website/docs/guides/migration/versions/upgrading-to-v1.2.md), describing the change at a high level (one sentence) and linking to this page?

callum-mcdata committed:
Update website/docs/docs/building-a-dbt-project/metrics.md

Co-authored-by: Jeremy Cohen <jeremy@dbtlabs.com>

callum-mcdata committed:
updating wording in metrics

callum-mcdata committed:
Merge branch 'expression_metrics_updates' of github.com:dbt-labs/docs.getdbt.com into expression_metrics_updates

jtcohen6 commented on website/docs/docs/building-a-dbt-project/metrics.md:
```diff
@@ -90,6 +92,85 @@ metrics:
+| Field       | Description                                                 | Example                         | Required? |
+|-------------|-------------------------------------------------------------|---------------------------------|-----------|
+| name        | A unique identifier for the metric                          | new_customers                   | yes       |
+| model       | The dbt model that powers this metric                       | dim_customers                   | <VersionBlock firstVersion="1.2">yes (no for `expression` metrics)</VersionBlock><VersionBlock lastVersion="1.1">yes</VersionBlock> |
```
Idea in making this change is that we can have just one table of "Available properties," rather than duplicating it here + below (beneath `<VersionBlock lastVersion="1.1">`)

callum-mcdata replied:
Accepted changes!

jtcohen6 replied:
Not quite. Here's a commit with what I mean: https://github.com/dbt-labs/docs.getdbt.com/commit/54b9b818db0121d2e8fea92690d169933585cdaf

jtcohen6 replied:
@callum-mcdata Still need to delete the duplicated version of this table below as well:

<img width="556" alt="Screenshot 2022-07-14 at 01 50 40" src="https://user-images.githubusercontent.com/13897643/178856008-38b1fdec-0472-4a83-abdf-2967fb35f377.png">

callum-mcdata replied:
Odd - when I spun it up on my local there was only one table 🤔  I wonder if I messed up on the merge conflict from current. Let me fix now!

callum-mcdata replied:
Yep that was it! Fixed it:

![image](https://user-images.githubusercontent.com/101437052/179008095-f88ce990-047d-44ce-8563-3dc5dfc4af73.png)

callum-mcdata committed:
reflecting changes in upgrade doc

callum-mcdata commented:
@jtcohen6 thanks for all the helpful comments! I've resolved the blocking issues and addressed the non-blocking issues (opened an issue for one of them).  The upgrading page has also been updated with the following line:

- Updated `metrics.md` to reflect the addition of the `metric` function and support for `expression` metrics.

Let me know if there is anything blocking for this PR!

jtcohen6 commented on website/docs/guides/migration/versions/upgrading-to-v1.2.md:
```diff
@@ -22,5 +22,6 @@ We added a collection of ["cross-database macros"](cross-database-macros) to dbt
 ## New and changed documentation
 
 _Under construction_
+- Updated `metrics.md` to reflect the addition of the `metric` function and support for `expression` metrics.
```
Two things:
- The intent here is not to keep a pure changelog, but rather to give users a "highlight reel" of what's changed, and links to read more
- You'll need to resolve a merge conflict before merging (most PRs for v1.2 target updates to this page)
```suggestion
- [**Metrics**](building-a-dbt-project/metrics) now support an `expression` type (metrics-on-metrics), as well as a `metric()` function to use when referencing metrics from within models or `expression`-type metrics.
```

jtcohen6 commented on website/docs/docs/building-a-dbt-project/metrics.md:
```diff
@@ -90,6 +92,85 @@ metrics:
+
+<VersionBlock firstVersion="1.2">
+
+### Available types
```
Any reason why we can't show this table (or at least most of it) to users of v1.0 + v1.1 as well? I'd propose removing the `VersionBlock` wrapping around the whole table, and then putting something like this for the last line:
```
| expression | <VersionBlock firstVersion="1.2">This metric type is defined as any **non-aggregating** calculation of 1 or more metrics</VersionBlock><VersionBlock lastVersion="1.1">Not yet available — added in v1.2</VersionBlock> |
```

callum-mcdata replied:
I'm on board with this idea - I think it was something missing from the pre-1.2 versions anyway so if I can backfill it that's a bonus in my book.

jtcohen6 reviewed (approved):
LGTM pending these two changes!
- [X] Update migration guide note (to resolve merge conflict)
- [X] DRY up "Available properties" table

**CM Edit**: *Tasks accomplished on 7/13/22*

callum-mcdata committed:
updating docs

callum-mcdata committed:
Merge branch 'current'

callum-mcdata committed:
fixing conflict

callum-mcdata committed:
remove extra properties

jtcohen6 reviewed (approved):
Let's ship it!

callum-mcdata commented:
@annafil @runleonarun I'll let y'all hit the big green button as you are the codeowners for this repo. Or if you'd prefer me to hit it, let me know!

jtcohen6 commented:
@callum-mcdata I think you're good to merge yourself, since you did the leg work here! But I'll press it if you don't want to ;)

callum-mcdata commented:
GONNA PRESS THAT BUTTON THEN! 🕺
